File: app/src/main/java/peter/util/searcher/TabManager.java
package peter.util.searcher;

import android.view.View;
import android.widget.Toast;

import java.util.ArrayList;

import peter.util.searcher.activity.MainActivity;
import peter.util.searcher.bean.Bean;
import peter.util.searcher.tab.SearcherTab;
import peter.util.searcher.tab.TabGroup;

/**
 * 标签管理类
 * Created by peter on 2016/11/17.
 */

public class TabManager {

    public static final int MAX_TAB = 50;
    private final ArrayList<TabGroup> tabGroupArrayList = new ArrayList<>(MAX_TAB);
    private final MainActivity mainActivity;
    private int mCurrentTabIndex;

    public TabManager(MainActivity activity) {
        mainActivity = activity;
    }

    public void loadUrl(Bean bean, boolean newTab) {
        if (tabGroupArrayList.size() == MAX_TAB) {
            newTab = false;
            Toast.makeText(Searcher.context, R.string.tabs_max_txt, Toast.LENGTH_LONG).show();
        }
        if (newTab) {
            TabGroup tab = new TabGroup(mainActivity);
            tabGroupArrayList.add(tab);
            mCurrentTabIndex = tabGroupArrayList.size() - 1;
        }
        TabGroup tabGroup = getCurrentTabGroup();
        tabGroup.loadUrl(bean);
    }

    public TabGroup getCurrentTabGroup() {
        return tabGroupArrayList.size() > 0 ? tabGroupArrayList.get(mCurrentTabIndex) : null;
    }

    public void switchTabGroup(TabGroup tabGroup) {
        switchTabGroup(tabGroup, true);
    }

    private void switchTabGroup(TabGroup tabGroup, boolean reload) {
        mCurrentTabIndex = tabGroupArrayList.indexOf(tabGroup);
        View currentTabGroupView = tabGroup.getCurrentTab().getView();
        pauseTabGroupExclude(tabGroup);
        mainActivity.setCurrentView(currentTabGroupView);
        getCurrentTabGroup().onResume();
        mainActivity.refreshTitle();
        if (reload) {
            getCurrentTabGroup().checkReloadCurrentTab();
        }
    }

    public void restoreTabPos(int groupIndex, int tabIndex) {
        TabGroup tabGroup = tabGroupArrayList.get(groupIndex);
        switchTabGroup(tabGroup, false);
        getCurrentTabGroup().setCurrentTab(tabIndex);
        getCurrentTabGroup().checkReloadCurrentTab();
    }

    public TabGroup getTabGroup(SearcherTab topTab) {
        for (TabGroup tabGroup : tabGroupArrayList) {
            if (tabGroup.containsTab(topTab)) {
                return tabGroup;
            }
        }
        return null;
    }

    public void removeTabGroup(TabGroup tabGroup) {
        if (tabGroupArrayList.size() > 0) {
            tabGroup.onDestroy();
            int removeIndex = tabGroupArrayList.indexOf(tabGroup);
            int nextIndex = 0;
            if(removeIndex < mCurrentTabIndex) {
                nextIndex = mCurrentTabIndex - 1;
            }else if(removeIndex > mCurrentTabIndex) {
                nextIndex = mCurrentTabIndex;
            }else if(removeIndex == mCurrentTabIndex) {
                if(removeIndex == tabGroupArrayList.size() - 1) {//end
                    nextIndex = tabGroupArrayList.size() - 2;
                }else {
                    nextIndex = mCurrentTabIndex;
                }
            }
            tabGroupArrayList.remove(tabGroup);
            if (tabGroupArrayList.size() > 0) {
                TabGroup tabGroupNext = tabGroupArrayList.get(nextIndex);
                switchTabGroup(tabGroupNext);
            }
        }
    }

    public void removeIndex(TabGroup tabGroup) {
        if (tabGroupArrayList.size() > 0) {
            tabGroupArrayList.remove(tabGroup);
        }
    }

    public void resumeTabGroupExclude(TabGroup exTabGroup) {
        ArrayList<TabGroup> list = new ArrayList<>(tabGroupArrayList);
        list.stream().filter(tabGroup -> tabGroup != exTabGroup).forEach(TabGroup::onResume);
    }

    public void pauseTabGroupExclude(TabGroup exTabGroup) {
        ArrayList<TabGroup> list = new ArrayList<>(tabGroupArrayList);
        list.stream().filter(tabGroup -> tabGroup != exTabGroup).forEach(TabGroup::onPause);
    }

    public ArrayList<TabGroup> getList() {
        return tabGroupArrayList;
    }

    public int getTabGroupCount() {
        return tabGroupArrayList.size();
    }

}

File: app/src/main/java/peter/util/searcher/TabManager.kt
package peter.util.searcher

import android.widget.Toast

import java.util.ArrayList

import peter.util.searcher.activity.MainActivity
import peter.util.searcher.bean.Bean
import peter.util.searcher.tab.SearcherTab
import peter.util.searcher.tab.TabGroup

/**
 * 标签管理类
 * Created by peter on 2016/11/17.
 */

class TabManager(private val mainActivity: MainActivity) {
    private val maxTabs = 50
    private var mCurrentTabIndex: Int = 0
    val list = ArrayList<TabGroup>(maxTabs)

    fun currentTabGroup(): TabGroup? = when {
        !list.isEmpty() -> list[mCurrentTabIndex]
        else -> null
    }

    fun loadUrl(bean: Bean, mNewTab: Boolean) {
        var newTab = mNewTab
        if (list.size == maxTabs) {
            newTab = false
            Toast.makeText(Searcher.context, R.string.tabs_max_txt, Toast.LENGTH_LONG).show()
        }
        if (newTab) {
            val tab = TabGroup(mainActivity)
            if (list.size == 0 || mCurrentTabIndex == list.size - 1) {//有0个tab 或 currentTab在最后
                mCurrentTabIndex = list.size
                list.add(tab)
            } else {//currentTab不在最后
                mCurrentTabIndex += 1
                list.add(mCurrentTabIndex, tab)
            }
        }
        val tabGroup = currentTabGroup()
        tabGroup?.loadUrl(bean)
    }

    fun switchTabGroup(tabGroup: TabGroup) {
        switchTabGroup(tabGroup, true)
    }

    private fun switchTabGroup(tabGroup: TabGroup, reload: Boolean) {
        mCurrentTabIndex = list.indexOf(tabGroup)
        val currentTabGroupView = tabGroup.currentTab()!!.getView()
        pauseTabGroupExclude(tabGroup)
        mainActivity.setCurrentView(currentTabGroupView!!)
        currentTabGroup()!!.onResume()
        mainActivity.refreshTitle()
        if (reload) {
            currentTabGroup()!!.checkReloadCurrentTab()
        }
    }

    fun restoreTabPos(groupIndex: Int, tabIndex: Int) {
        val tabGroup = list[groupIndex]
        switchTabGroup(tabGroup, false)
        currentTabGroup()!!.setCurrentTab(tabIndex)
        currentTabGroup()!!.checkReloadCurrentTab()
    }

    fun getTabGroup(topTab: SearcherTab): TabGroup? = list.firstOrNull { it.containsTab(topTab) }

    fun removeTabGroup(tabGroup: TabGroup) {
        if (list.size > 0) {
            tabGroup.onDestroy()
            list.remove(tabGroup)
            if (list.size > 0) {
                val indexNext = list.size - 1
                val tabGroupNext = list[indexNext]
                switchTabGroup(tabGroupNext)
            }
        }
    }

    fun removeIndex(tabGroup: TabGroup) {
        if (list.size > 0) {
            list.remove(tabGroup)
        }
    }

    fun resumeTabGroupExclude(exTabGroup: TabGroup?) {
        list.filter { it !== exTabGroup }.forEach { it.onResume() }
    }

    fun pauseTabGroupExclude(exTabGroup: TabGroup?) {
        list.filter { it !== exTabGroup }.forEach { it.onPause() }
    }

}
